designdocs/permissions.md

```diff
@@ -0,0 +1,87 @@
+
+Concerns about both systems include:
+* the user agents know more about the user, the current scenario, and can make better decisions *at the time of use* than a spec can. Neither method give the user agent freedom to adapt.
+* there may be new attacks devised or new flaws discovered after we have shipped, but user agents will not be able to address these because they are constrained by the spec.
+
+There are most likely more concerns than just what I have listed above.
+
+### Solution
+To address these I propose the following: *do both*.  Let the developers have a single code path where they /request permissions at application start/ *and* when /the permission is actually needed/. The first request is considered the upper bounds of possible permissions. The user agent can decide whether to actually show user request dialogs at app time or when the permission is actually used, or some other scheme that the user agent deems to be better (such as warding off new attacks in the future).
+
+### Details
+From the developer’s perspective it would work like this.
+
+The developer specifies the total permissions they are likely to need during the entire run of the application when requesting the XR Session, using an array with pre-defined permissions constants.  Something like this:
+
+``` javascript
+const perms = [
+   MICROPHONE,
+	 IMMERSIVE_MODE,
+   HEADSET_POSE,
+   //CAMERA_DATA,  developer decides not to use camera
+]
+navigator.xr.requestSession(perms).then(session => {
+	//start up the rest of the app
+})
+```
+
+At this point the user agent can either prompt the user for these permissions, or merely use them as an upper bound of permissions that could potentially be requested later.
+
+
+Later, when the application needs to actually use a permission, say the microphone, the developer will request this access through a promise based API. Ex:
+
+```javascript
+navigator.mediaDevices.getUserMedia(opts).then(stream => {
+	console.log("i have the microphone stream")
+})
+```
+
+At this point the user agent can either prompt the user for the microphone permission, or immediately grant it if the user already approved it previously.  
+
+Note that if the application did not include MICROPHONE as  the initial list of permissions to `requestSession` then the user agent *must* reject any later requests to `getUserMedia()`.  
+
+*The initial list of permissions is the maximum set of permissions the application may request throughout it’s lifetime.*  This enables, say, a vr chat client to let the user enter a move and talk only mode and later request camera access because the user wants to share something, without having to bug the user initially; if the user agent supports immersive permission requests.  On user agents which do not support immersive permission requests, for whatever reason (form factor, user prefs, domain name, phase of moon, etc.), then the request for camera can be done at application start.
```

Comment:
"lifetime" is an important concept here - do you want to specifically say "duration of the session"?

If I understand it correctly, the application would be free to end the session and start a new one with a different set of desired permissions, while still preserving JS and graphics state. That has its own issues, i.e. a desktop VR setup may not offer an easy way to resume the session, but I think it's a potential alternative for developers. An app could have distinct modes where some features are only needed in specific circumstances, for example a VR sculpting app with optional copresence support?

Reply:
Yeah, this is totally a key thing we need to nail down..  From what I'm gathering there are 3 possibilities under consideration for the lifetime of a permission:
* Session
* Browsing context
* Origin

The big question I have is... can this be something left to the user agent to decide?

Reply:
Since these permissions are requested when obtaining the session, I think they should be for the lifetime of the session.

Reply:
The [privacy design doc](https://github.com/immersive-web/webxr/pull/638) suggests that for a given origin, user consent for a set of feature should last as long as the browsing context. This means the same origin doesn't have to repeatedly ask for user consent for the same set of features during the same browsing session. However, if the user closes the browser and restarts, then user consent would be required.